### PR TITLE
scripts: zephyr modules: introducing Zephyr module dependency handling

### PR DESCRIPTION
**Pull request by tejlmand**
This commit introduces the possibility of adding dependencies between
Zephyr modules to ensure that a Zephyr module on which other modules
depends is processed first.

This is configured in the `zephyr/module.yml` file, as example:
```
build:
  cmake: .
  depends:
    - fatfs
```

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

File: scripts/zephyr_module.py [hunk 1]
```diff
@@ -215,20 +220,52 @@ def main():
...
+            deps = section.get('depends', [])
+            if not deps:
+                start_modules.append( [project, meta] )
+            else:
+                dep_modules.append( [project, meta, deps] )
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    # This will do a topology search to ensure the modules are ordered
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:244]: "topological" please

File: scripts/zephyr_module.py [hunk 2]
```diff
@@ -215,20 +220,52 @@ def main():
...
+    # according to dependency settings.
+    while start_modules:
+        # We want to preserve order as close to that returned by west.
+        # Thus remove elements from beginning, instead of pop().
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:248]: `node = start_modules.pop(0)`

File: scripts/zephyr_module.py [hunk 3]
```diff
@@ -215,20 +220,52 @@ def main():
...
 
         meta = process_module(project)
         if meta:
-            kconfig += process_kconfig(project, meta)
-            cmake += process_cmake(project, meta)
-            sanitycheck += process_sanitycheck(project, meta)
+            section = meta.get('build', dict())
+            deps = section.get('depends', [])
+            if not deps:
+                start_modules.append( [project, meta] )
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:237]: Please fix whitespace nits here and elsewhere, removing spaces between ( and [, as well as between ] and ).

File: scripts/zephyr_module.py [hunk 4]
```diff
@@ -215,20 +220,52 @@ def main():
...
+    # according to dependency settings.
+    while start_modules:
+        # We want to preserve order as close to that returned by west.
+        # Thus remove elements from beginning, instead of pop().
+        node = start_modules[0]
+        del start_modules[0]
+        sorted_modules.append(node)
+        node_name = PurePath(node[0]).name
+        for module in dep_modules:
+            if node_name in module[2]:
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:254]: Would you mind using a collections.namedtuple as a data structure just to make this more readable? I don't remember how this script works, and am feeling very lazy about chasing down what the the magic number indexes mean. It's okay if you don't have time; I'll make a better effort if that's the case :).

File: scripts/zephyr_module.py [hunk 5]
```diff
@@ -215,20 +220,52 @@ def main():
...
+                module[2].remove(node_name)
+                if not module[2]:
+                    start_modules.append(module)
+                    dep_modules.remove(module)
+
+    if dep_modules:
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:260]: Could you please document or explain more clearly what happens when a dependency is missing? I expected to see an error somewhere and it's not popping out at me.

**Comment by mbolivar-nordic** [comment]
100% on board with the idea, just would like to see more details on the semantics here.

File: scripts/zephyr_module.py (hunk 1 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:244]: and not search, should be sort.
I hope I do better coding than commenting when it gets late.

File: scripts/zephyr_module.py (hunk 5 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:260]: This is no a dependency missing, but if there is still entries in `dep_modules` and processing of `start_modules` has been completed, it means that the remaining modules has a cyclic dependency.

Adding a dependency is not required, but if a module is dependent on another module, that module must be earlier in the sorted list.

File: scripts/zephyr_module.py (hunk 4 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:254]: good input, will clean up.
I was just too focused on getting things up and running, and still having a bit of C habits when writing python. Need to get out of dinosaur mode.

File: scripts/zephyr_module.py (hunk 3 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:237]: Will do.
Maybe it's my eyes, or I'm using a too small font size, but I always have hard time reading `([text])` as the parenthesis and square brackets kind of float together, thus I find an extra space easing the reading.

File: scripts/zephyr_module.py (hunk 2 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:248]: Why have I always been under the impression I could only `pop()` or `pop(object)`, and not based on index. 🤔

File: scripts/zephyr_module.py (hunk 4 above)
Comment (tejlmand) [reply scripts/zephyr_module.py]: Transition from dinosaur mode completed.

File: scripts/zephyr_module.py (hunk 5 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:264]: added some comments.

File: scripts/zephyr_module.py (hunk 1 above)
Comment (tejlmand) [reply scripts/zephyr_module.py]: fixed

File: scripts/zephyr_module.py (hunk 2 above)
Comment (tejlmand) [reply scripts/zephyr_module.py]: fixed

File: scripts/zephyr_module.py (hunk 3 above)
Comment (tejlmand) [reply scripts/zephyr_module.py]: And changing to tuple helped with the `([text])` as well.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, but how/where are unmet dependencies handled? I see handling for cyclic dependencies, but no explicit handling/error for unmet dependencies.

**Comment by tejlmand** [comment]
> Looks good, but how/where are unmet dependencies handled? I see handling for cyclic dependencies, but no explicit handling/error for unmet dependencies.

An unmet dependency like:
```
build:
  cmake: .
  depends:
    - invalid_project
```
will never get removed from `dep_modules`, and as such show up at the final `if dep_modules:`.

Will adjust error description.

**Review by henrikbrixandersen — APPROVED** [review]
LGTM.

File: scripts/zephyr_module.py (hunk 5 above)
Comment (mbolivar-nordic) [reply scripts/zephyr_module.py:265]: Is it possible to separate the cases of cyclic dependencies versus unmet dependencies? I think unmet dependencies should be relatively easy to filter out. The rest would be cyclic. I think the output would be more readable if the different cases produced different output personally, but I won't NAK over it.

File: scripts/zephyr_module.py [hunk 6]
```diff
@@ -215,20 +221,58 @@ def main():
...
+                if not module.depends:
+                    start_modules.append(module)
+                    to_remove.append(module)
+        for module in to_remove:
+            dep_modules.remove(module)
+
+    if dep_modules:
+        # If there are any modules with unresolved dependencies, then the
+        # modules contains unmet or cyclic dependencies. Error out.
+        modules = [ x.project for x in dep_modules ]
```
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:268]: I imagine the linter will complain about the spaces after [ and before ]? If it doesn't, no worries.

**Review by mbolivar-nordic — APPROVED** [review]
Looks fine, though I would hope to see the error handling improved at some point.

File: scripts/zephyr_module.py (hunk 6 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:268]: Fixed now, eagle eye.

File: scripts/zephyr_module.py (hunk 5 above)
Comment (tejlmand) [reply scripts/zephyr_module.py:265]: Now printing:
```
CMake Error at /projects/github/ncs/zephyr/cmake/zephyr_module.cmake:45 (message):
  Unmet or cyclic dependencies in modules:

  /projects/github/ncs/modules/hal/cmsis depends on: ['bda_splelnig']

  /projects/github/ncs/modules/fs/fatfs depends on: ['nordic']

  /projects/github/ncs/modules/hal/nordic depends on: ['fatfs']
```
don't like two split the two, cause there could be both unmet deps AND cyclic deps at the same time.

But now the deps are printed a long with the modules, which should make it very easy to find the root cause of the failing dependency.

Comment (mbolivar-nordic) [reply scripts/zephyr_module.py:265]: Perfect. Thanks.

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**